### ISSUE TITLE
NPE in Refactoring

### ISSUE THREAD

**Issue by cdietrich** [eclipse/xtext-core#300]
Rename Refactoring the Bar class in following code

```
class Foo {
	new(Bar22222344 bar) {
	}
}

class Bar22222344 {
	val foo1 = new Foo(this) {
	}
	
	val foo2 = new Foo(this) {
	}
		
	val foo3 = new Foo(this) {
	}
}
```
i sometimes get this NPE

```
197987 [main] ERROR org.eclipse.xtext.ui.refactoring.ui.DefaultRenameElementHandler  - Error initializing refactoring
java.lang.NullPointerException
	at org.eclipse.xtext.parsetree.reconstr.impl.DefaultTransientValueService.isTransient(DefaultTransientValueService.java:24)
	at org.eclipse.xtext.serializer.sequencer.LegacyTransientValueService.isValueInListTransient(LegacyTransientValueService.java:36)
	at org.eclipse.xtext.ui.refactoring.impl.DefaultReferenceUpdater.createReferenceUpdate(DefaultReferenceUpdater.java:120)
	at org.eclipse.xtext.xbase.ui.jvmmodel.refactoring.JvmModelReferenceUpdater.createReferenceUpdate(JvmModelReferenceUpdater.java:50)
	at org.eclipse.xtext.xbase.ui.refactoring.XbaseReferenceUpdater.createReferenceUpdate(XbaseReferenceUpdater.java:155)
	at org.eclipse.xtext.ui.refactoring.impl.DefaultReferenceUpdater.createReferenceUpdate(DefaultReferenceUpdater.java:114)
	at org.eclipse.xtext.ui.refactoring.impl.DefaultReferenceUpdater.processReferringResource(DefaultReferenceUpdater.java:93)
	at org.eclipse.xtext.xbase.ui.refactoring.XbaseReferenceUpdater.processReferringResource(XbaseReferenceUpdater.java:237)
	at org.eclipse.xtext.ui.refactoring.impl.DefaultReferenceUpdater.createReferenceUpdates(DefaultReferenceUpdater.java:78)
	at org.eclipse.xtext.ui.refactoring.impl.AbstractReferenceUpdater.createReferenceUpdatesForCluster(AbstractReferenceUpdater.java:152)
	at org.eclipse.xtext.ui.refactoring.impl.AbstractReferenceUpdater.createClusteredReferenceUpdates(AbstractReferenceUpdater.java:104)
	at org.eclipse.xtext.ui.refactoring.impl.AbstractReferenceUpdater.createReferenceUpdates(AbstractReferenceUpdater.java:80)
	at org.eclipse.xtext.ui.refactoring.impl.AbstractReferenceUpdater.createReferenceUpdates(AbstractReferenceUpdater.java:59)
	at org.eclipse.xtext.ui.refactoring.ui.DefaultLinkedPositionGroupCalculator.getLinkedPositionGroup(DefaultLinkedPositionGroupCalculator.java:174)
	at org.eclipse.xtext.ui.refactoring.ui.RenameRefactoringController$1.run(RenameRefactoringController.java:145)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:119)

```

reason is the referringElement calculated in `org.eclipse.xtext.ui.refactoring.impl.DefaultReferenceUpdater.createReferenceUpdate(IReferenceDescription, URI, ElementRenameArguments, ResourceSet, IRefactoringUpdateAcceptor)`
yields null.

i dont know why this is the case.

the uri that is searched for has a fragment like

```
/2/@members.2/@localClasses.0/@members.0/@parameters.0/@parameterType
```
the local class has i  this case no member

=> this might be a problem in xtend as well

**Comment by cdietrich** [comment]
found this while investigating https://bugs.eclipse.org/bugs/show_bug.cgi?id=489208

**Comment by cdietrich** [comment]
i propose to simply add a null check to the updater.
@JanKoehnlein what do you think

**Comment by cdietrich** [comment]
org.eclipse.xtend.ide.tests.refactoring.JavaRefactoringIntegrationTest.testBug489208_1() and
org.eclipse.xtend.ide.tests.refactoring.JavaRefactoringIntegrationTest.testBug489208_2() sometimes run into this problem

**Comment by cdietrich** [comment]
(the totel set of ref. desc is the same in both cases) => it seems the reordering makes it bad

**Comment by cdietrich** [comment]
potentialy bogus places: 
- org.eclipse.xtext.ui.refactoring.impl.ReferenceDescriptionSorter (should use LinkedHashMultimap)
- org.eclipse.xtext.ui.refactoring.impl.ReferenceUpdaterDispatcher.ReferenceDescriptionAcceptor (same)

@JanKoehnlein what do you think?

**Comment by JanKoehnlein** [comment]
I'd expect a sorter to return elements in a deterministic order, but the `ReferenceDescriptionSorter` states in its JavaDoc that it only sorts on resource / project level.

It would be more interesting to find out why the order matters, as the reference updating should not depend on it.

**Comment by cdietrich** [comment]
i assume the calling of resource.getEObject(...) affects the inferrer/inner class creation

**Comment by cdietrich** [comment]
(there was still the point of why there where references for the inner class at all)

**Comment by cdietrich** [comment]
hmmm sorting actually makes the localclasses come before the members, which makes the test fail 100% of time

**Comment by cdietrich** [comment]
org.eclipse.xtend.core.typesystem.XtendReentrantTypeResolver.inferAnonymousClassConstructor(AnonymousClass, JvmGenericType, JvmConstructor, boolean)

creates the "bad" constructor.

i wonder why this not always happens

**Comment by cdietrich** [comment]
trace in the "good" case

```
XtendReentrantTypeResolver(LogicalContainerAwareReentrantTypeResolver).computeTypes(ResolvedTypes, IFeatureScopeSession) line: 698	
XtendReentrantTypeResolver.computeTypes(ResolvedTypes, IFeatureScopeSession) line: 382	
XtendReentrantTypeResolver(DefaultReentrantTypeResolver).resolve(CancelIndicator) line: 163	
XtendReentrantTypeResolver(DefaultReentrantTypeResolver).reentrantResolve(CancelIndicator) line: 139	
CompoundReentrantTypeResolver.reentrantResolve(CancelIndicator) line: 79	
CachingBatchTypeResolver$LazyResolvedTypes.resolveTypes(CancelIndicator) line: 80	
CachingBatchTypeResolver$2.process(Resource) line: 57	
CachingBatchTypeResolver$2.process(Object) line: 1	
CachingBatchTypeResolver$2(IUnitOfWork$Void<T>).exec(T) line: 37	
OnChangeEvictingCache.execWithoutCacheClear(Param, IUnitOfWork<Result,Param>) line: 129	
CachingBatchTypeResolver.doResolveTypes(EObject, CancelIndicator) line: 53	
CachingBatchTypeResolver(AbstractBatchTypeResolver).resolveTypes(EObject, CancelIndicator) line: 69	
CachingBatchTypeResolver(AbstractBatchTypeResolver).resolveTypes(EObject) line: 55	
XtendScopeProvider(XbaseBatchScopeProvider).getScope(EObject, EReference) line: 101	
XtendScopeProvider(XbaseWithAnnotationsBatchScopeProvider).getScope(EObject, EReference) line: 53	
JvmModelReferenceUpdater$1.isValid(IEObjectDescription) line: 111	
RefactoringCrossReferenceSerializer.getCrossRefText(EObject, CrossReference, EObject, RefactoringCrossReferenceSerializer$RefTextEvaluator, ITextRegion, StatusWrapper) line: 74	
XtendReferenceUpdater(DefaultReferenceUpdater).createReferenceUpdate(EObject, URI, EReference, int, EObject, IRefactoringUpdateAcceptor) line: 126	
XtendReferenceUpdater(JvmModelReferenceUpdater).createReferenceUpdate(EObject, URI, EReference, int, EObject, IRefactoringUpdateAcceptor) line: 50	
```

@JanKoehnlein i assume at this point everthing shall already be resolved and members be in place? but they are not, after the call they are => this is why it depends on order

**Comment by cdietrich** [comment]
marked tests in JavaRefactoringIntegrationTest as flaky

**Comment by cdietrich** [comment]
fixed in 2.12
